### ISSUE TITLE
The `vaadinPrepareFrontendTask` gradle task is not cached

### ISSUE THREAD

**Issue by steve-todorov** [vaadin/flow#17941]
### Description of the bug

We have upgrade to Vaadin `24.2.0` with the hopes to benefit from the newly introduced incremental Gradle task caching for `vaadinPrepareFrontendTask` (related PR is https://github.com/vaadin/flow/pull/17150). 

Unfortunately our local and CI builds seem to be unable to benefit from this feature.
Upon closer investigation we noticed Gradle scans show the following notice:

![image](https://github.com/vaadin/flow/assets/2303663/15c665ab-5480-4be7-a5f6-543ff16b541c)

### Expected behavior

The `vaadinPrepareFrontendTask` should show `UP-TO-DATE` or `FROM CACHE` in the gradle build scans and the task should not be computing again.

### Minimal reproducible example

Not sure how to reproduce yet, but we don't have any special setup.
We are using the `gradle vaadin plugin` and run `gradle build`.

### Versions

- Vaadin / Flow version: 24.2.0
- Java version: 17
- OS version: ubuntu-22.04 (GH CI)
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

**Comment by mshabarov** [comment]
Thanks for the issue. We will try to reproduce with our Gradle-based starters and get back to you soon.

**Comment by steve-todorov** [comment]
Thanks for looking into this @mshabarov! 

Let me know if you need me to give you additional context or if you need anything more.

**Comment by tepi** [comment]
Tested this out with Vaadin 24.2.0 and the [base starter](https://github.com/vaadin/base-starter-gradle). There the caching seems to work fine. This is the output I get:

**Comment by steve-todorov** [comment]
Hey @tepi,

The `base starter` project does not appear to have a `package.json` and `package-lock.json` files.
In our project we have both of them. My current working theory is the `vaadinBuildFrontend` task might be corrupting the `cache`, because sometimes it updates packages in the `package.json` which were not committed into the git.

![image](https://github.com/vaadin/flow/assets/2303663/d0b3f95d-8c48-46bf-b673-4176131fd6f4)

So what appears to be happening is that the `vaadinBuildFrontend` is overriding the `package.json` file and thus changing the cache key that was initially generated by the `vaadinPrepareFrontend` task. I have been able to reproduce the problem with the base starter project.

You can have a look in my fork located here: 
https://github.com/steve-todorov/base-starter-gradle/tree/issue-17941

Also checkout the GH actions here:
https://github.com/steve-todorov/base-starter-gradle/actions/runs/6724319472

The [Gradle build scan](https://scans.gradle.com/s/ukovbnymd2jom/timeline?details=afz2wjxxnx4ms&hide-timeline) shows the same problem:

![image](https://github.com/vaadin/flow/assets/2303663/2740b739-1b3d-4b56-9f1e-d805c293be1d)

The way to reproduce it is by doing the following steps:

1. `rm -rf ~/.gradle/caches/build-cache-1/`
3. `gradle clean`
2. Have your `package.json` and `package-lock.json` committed in VCS
3. Your `package.json` has one package that is outdated (check the `package.json` --- the `@vaadin/upload": "24.1.0` should be `24.2.1`)
5. `gradle build`

**Comment by tepi** [comment]
Most likely without knowing all th relevant context here, the caching seems to work like it should. `vaadinPrepareFrontend` will update dev dependencies only and `vaadinBuildFrontend` will update frontend dependencies to the `package.json` file. If there is an outdated frontend dependency or other discrepancy between `package.json` and `package-lock.json` it should be updated by `vaadinBuildFrontend` task - and at this point the cache is no longer valid. Once the files are up-to-date, the updates should be committed to VCS, after which caching should work on subsequent builds.

If it is Vaadin that causes the outdated dependency to exist after a successfull build, that is probably a separate issue.

**Comment by mshabarov** [comment]
It's maybe fine if `vaadinBuildFrontend` overrides `package.json` within first build or after changes in the project. However, it sounds incorrect If `vaadinBuildFrontend` overrides `package.json` within subsequent builds, when no changes are applied to the project.

I propose to focus on why `vaadinBuildFrontend` updates it.

We can test and see what is changed by manually running `vaadinPrepareFrontend`, saving `package.json` and then running `vaadinBuildFrontend` and seeing what it changes.

**Comment by tepi** [comment]
With the provided example run locally, `vaadinBuildFrontend` updates `package.json` once (the outdated frontend dependency) and after that no more updates.

**Comment by mshabarov** [comment]
@steve-todorov you mentioned that `"@vaadin/upload": "24.1.0"` is updated to 24.2.1 during the `vaadinBuildFrontend`. This is for the first run, right? What is updated in `package.json` by `vaadinBuildFrontend` in subsequent runs, can you tell?

**Comment by oliveryasuna** [comment]
@mshabarov You know how sometimes the plugin's version is sometimes behind Vaadin Platform's. At least I remember this in the Maven plugin. Both the plugin and startup code use shared code in `vaadin-server` that reads `com/vaadin/flow/server/frontend/dependencies/default/package.json` on the classpath. If the plugin's version is behind the dependency's, prepare frontend will produce a different `package.json` then after running the app in dev mode. Then, if you ran build frontend, it would change again, being closer to a run's `package.json`, but may be different because build frontend does more stuff (I forget exactly at the moment).

Perhaps something related to this is the problem.

**Comment by oliveryasuna** [comment]
Additional thoughts... Not a solution but definitely part of a stable one.

Disclaimer: I am not a Gradle expert. Could you leverage the power of Gradle to make the `vaadinPrepareFrontend` and `vaadinBuildFrontend` tasks always priority your application's classpath, thus ensuring that `/flow/server/frontend/dependencies/default/package.json` from the Vaadin Platform version you have as a dependency is considered? @steve-todorov?

There is a serious drawback to this approach. Gradle configurations can be complex. I have seen situations where an addon uses a different Vaadin Platform version and ends up taking priority of the Vaadin Platform version used. But for some weird reason, when the app runs, it uses the right `/flow/server/frontend/dependencies/default/package.json`.

**Comment by knoobie** [comment]
> You know how sometimes the plugin's version is sometimes behind Vaadin Platform's.

That sounds like a bug in your build. Both should always be the same.

**Comment by oliveryasuna** [comment]
> > You know how sometimes the plugin's version is sometimes behind Vaadin Platform's.
> 
> That sounds like a bug in your build. Both should always be the same.

I was not clear. They used to release the Maven plugin a few days or sometimes weeks after Platform. Not sure if they still do this and it doesn't matter for us since switching it Gradle it seems they release the Gradle plugin at the same time.

**Comment by knoobie** [comment]
Since 2x where the Gradle plugin was officially adopted, this shouldn't be a problem anymore. Now the release process looks simplified like this: 
1. Flow including Spring, Maven and Gradle (and others) 
2. Platform release with anything Vaadin related is build and tested together including the flow version of 1.

**Comment by mvysny** [comment]
> Additional thoughts... Not a solution but definitely part of a stable one.
> 
> Disclaimer: I am not a Gradle expert. Could you leverage the power of Gradle to make the `vaadinPrepareFrontend` and `vaadinBuildFrontend` tasks always priority your application's classpath, thus ensuring that `/flow/server/frontend/dependencies/default/package.json` from the Vaadin Platform version you have as a dependency is considered? @steve-todorov?
> 
> There is a serious drawback to this approach. Gradle configurations can be complex. I have seen situations where an addon uses a different Vaadin Platform version and ends up taking priority of the Vaadin Platform version used during build. But for some weird reason, when the app runs in dev mode, it uses the right `/flow/server/frontend/dependencies/default/package.json`. I have also seen this reversed.

I'm not sure I fully understand the issue. This sounds unrelated to the original problem - could you please create a new ticket for this, so that we can discuss this in deeper detail?

**Comment by mvysny** [comment]
@oliveryasuna it's very important to always use the same version of the Vaadin Gradle Plugin and the Vaadin itself, otherwise the Plugin will use slightly different version of flow-server.jar on its classpath, which could cause random bugs to appear.

Also, it is possible for `vaadinBuildFrontend` to modify `package.json` once, after a new dependency has been added or Vaadin has been upgraded. If that happens, `vaadinPrepareFrontend` cache will be considered expired and it's normal to see it being run fully. If `vaadinBuildFrontend` constantly keeps overwriting/modifying the `package.json`, then that's definitely a bug which needs to be reported and resolved. You can try to reproduce such a bug by running full build `vaadinBuildFrontend` in production mode once on your dev machine, then commit `package.json` to git, then run the same build on other machines - the `package.json` should not be modified by consequent builds, and ultimately the `vaadinPrepareFrontend` task should be cached properly.

**Comment by mvysny** [comment]
If I understand this correctly: the change to the `package.json` file is causing `vaadinPrepareFrontend` task cache to become stale, to which Gradle correctly responds by re-running `vaadinPrepareFrontend` in full. This is a correct behavior. The excessive modification of `package.json` file by `vaadinBuildFrontend` sounds like a bug, but I believe that to be a bug different from this one, and a new bug report should be opened.

Therefore, if everyone agrees, I propose to close this bug as "works as intended". @mshabarov @oliveryasuna @knoobie

**Comment by mshabarov** [comment]
Thanks @mvysny @oliveryasuna @knoobie for you analysis!
I like the idea to open a new one, but before I would like to understand what changes in `package.json` are made by `vaadinBuildFrontend` for @steve-todorov .

**Comment by steve-todorov** [comment]
Hey guys,

Apologies for the late reply.

> @mvysny If vaadinBuildFrontend constantly keeps overwriting/modifying the package.json, then that's definitely a bug which needs to be reported as a separate issue and resolved by Vaadin. You can try to reproduce such a bug by running full build vaadinBuildFrontend in production mode once on your dev machine, then commit package.json to git, then run the same build on other machines - the package.json should not be modified by consequent builds, and ultimately the vaadinPrepareFrontend task should be cached properly.

What you're described is exactly what's happening. The original problem is that the `package.json` is overwritten during the CI build and then for some reason it's never able to be cached. Even if the `package.json` and `package-lock.json` are synced and committed to VCS the `vaadinPrepareFrontend` task is still not cached. We also tried clearing the GH Actions cache and running a few runs on the `main` but the result remains the same.

I am not entirely sure where this is coming from exactly, but the closest I could get to reproducing the problem was by introducing out-of-sync in the `package.json` as shown in https://github.com/steve-todorov/base-starter-gradle/tree/issue-17941. This leads me to believe something additional is happening in the `vaadinBuildFrontend` task that might be changing the `package.json` or `package-lock.json` causing the problem.   

I'll record the `package.json` and `package-lock.json` during the CI and diff them in hopes to get more clarity and report back.

**Comment by steve-todorov** [comment]
Hey @mshabarov, @mvysny, @tepi 

After some additional investigation we found out we had a few `@NpmPackage` annotations in our code which were using slightly older versions of `@vaadin` scoped packages. This was causing the issues with the `package.json` and `package-lock.json`.

When we removed those annotations and versions are now in sync (and in VCS).
This fixed our local cache, but it still isn't working in GH Actions.

Here's an example: 
https://github.com/steve-todorov/base-starter-gradle/actions/runs/6830298961

Build scan:
https://scans.gradle.com/s/2ya5bzgbshtms/timeline?details=afz2wjxxnx4ms&hide-timeline

Maybe we need to have some additional files cached?

**Comment by tepi** [comment]
Thanks for the additional info @steve-todorov. It is at least progress that the local cache works now. I checked out your branch and can confirm that. So issue could be something in GH Actions runner or config as well.

Do you know if there's any chance of getting a diff on what actually changes in package.json during the GH action run?

Also, could you try out running `clean vaadinClean build` instead of `clean build` in GHA and see if that makes any difference?

**Comment by steve-todorov** [comment]
In our project nothing really changes in the `package.json` nor `package-lock.json` - the md5 checksums stay the same so they should be binary identical.

I've updated my base-starter-gradle branch to include your suggestion to have `vaadinClean` in the mix, but that simply removes the `package-lock.json`:

https://github.com/steve-todorov/base-starter-gradle/actions/runs/6849622432

Build scan (continues to show `Overlapping outputs: Gradle does not know how file 'package.json' was created`):
https://scans.gradle.com/s/vj756oofsmo7s/timeline?details=afz2wjxxnx4ms&hide-timeline

**Comment by mvysny** [comment]
This is really strange. Only the `vaadinPrepareFrontend` task declares its input/output parameters (the `taskOutputProperties.packageJson` output parameters); `vaadinBuildFrontend` hasn't been converted to take advantage of the Gradle caching mechanism yet, and therefore it doesn't declare its inputs/outputs. Despite that, Gradle is able to detect the "overlapping outputs": that something else than `vaadinPrepareFrontend` modified the `package.json` file.

My hunch/wild guess is as following:
1. `vaadinBuildFrontend` overwrites `package.json` with the same contents, which modifies the last-access timestamp.
2. This last-access timestamp is taken into consideration by Gradle (for some reason only in the GHA environment).
3. Gradle detects a 'change' in `package.json` file, disables the build caching and warns about "overlapping outputs".

If we could somehow disable `vaadinBuildFrontend` touching/overwriting `package.json` completely, then we could put this hypothesis into test. @tepi  would you happen to know whether it's possible to do this?

**Comment by tepi** [comment]
Can confirm point 1 from @mvysny above; the timestamp does update even though the contents do not change. This might be the underlying reason for this behavior. Unfortunately could not find any way to disable this update - rather I consider it a bug that needs to be fixed, since it makes no sense to write the file if the contents are identical.

**Comment by mvysny** [comment]
Sounds good! In such case a new bug ticket should be opened in this repository, for the Gradle plugin. @steve-todorov could you take care of it please?

**Comment by steve-todorov** [comment]
@mvysny, @tepi I'm having a hard time understanding how points 1,2,3 would not work only in GH Actions though? It seems to be working fine when I'm doing local builds. If this were true I would expect the same behavior to appear locally?

**Comment by mvysny** [comment]
@steve-todorov maybe your local FS has disabled the tracking of last-update timestamp while it's enabled in the GHA environment... It's a bit far-fetched, but that's the only hypothesis I have.

**Comment by steve-todorov** [comment]
@mvysny I'm using Ubuntu 23.04 with ext4 which is similar to what the default GH Actions runner are using with docker. I did a small experiment to validate your theory. Locally the `package.json` and `package-lock.json` do not change as can be seen below

<details><summary>Local build, Cold cache (meaning no cache)</summary>

```
    steve@steve:base-starter-gradle$ stat package.json 
      File: package.json
      Size: 9926      	Blocks: 24         IO Block: 4096   regular file
    Device: 253,0	Inode: 831172      Links: 1
    Access: (0664/-rw-rw-r--)  Uid: ( 1000/   steve)   Gid: ( 1000/   steve)
    Access: 2023-11-13 16:39:47.348239724 +0200
    Modify: 2023-11-06 17:49:47.152887023 +0200
    Change: 2023-11-06 17:49:47.152887023 +0200
     Birth: 2023-11-01 21:42:01.779461450 +0200
    
    steve@steve:base-starter-gradle$ stat package-lock.json 
      File: package-lock.json
      Size: 576103    	Blocks: 1128       IO Block: 4096   regular file
    Device: 253,0	Inode: 794593      Links: 1
    Access: (0664/-rw-rw-r--)  Uid: ( 1000/   steve)   Gid: ( 1000/   steve)
    Access: 2023-11-13 16:39:47.430239538 +0200
    Modify: 2023-11-06 17:49:51.054877668 +0200
    Change: 2023-11-06 17:49:51.054877668 +0200
     Birth: 2023-11-06 17:48:47.668030020 +0200
```

</details>

<details><summary>Local build, Warm cache</summary>
    
    steve@steve:base-starter-gradle$ stat package.json
      File: package.json
      Size: 9926      	Blocks: 24         IO Block: 4096   regular file
    Device: 253,0	Inode: 831172      Links: 1
    Access: (0664/-rw-rw-r--)  Uid: ( 1000/   steve)   Gid: ( 1000/   steve)
    Access: 2023-11-13 16:39:47.348239724 +0200
    Modify: 2023-11-06 17:49:47.152887023 +0200
    Change: 2023-11-06 17:49:47.152887023 +0200
     Birth: 2023-11-01 21:42:01.779461450 +0200
    
    steve@steve:base-starter-gradle$ stat package-lock.json
      File: package-lock.json
      Size: 576103    	Blocks: 1128       IO Block: 4096   regular file
    Device: 253,0	Inode: 794593      Links: 1
    Access: (0664/-rw-rw-r--)  Uid: ( 1000/   steve)   Gid: ( 1000/   steve)
    Access: 2023-11-13 16:39:47.430239538 +0200
    Modify: 2023-11-06 17:49:51.054877668 +0200
    Change: 2023-11-06 17:49:51.054877668 +0200
     Birth: 2023-11-06 17:48:47.668030020 +0200

</details>

However, the build in GH does indeed `modify` the `package-lock.json` which might validate your theory:

https://github.com/steve-todorov/base-starter-gradle/actions/runs/6852381698

**Comment by oliveryasuna** [comment]
@mvysny I am a bit confused. Are you waiting on @steve-todorov to open a new issue? Why?

**Comment by tepi** [comment]
I'm looking into the issue currently, no need to create a new issue, I will create one when needed. Right now I can't reliably reproduce the timestamp update of `package.json` file. With identical setup it sometimes updates and sometimes does not update for my local build. Will keep you posted when I figure out what is going on.

**Comment by oliveryasuna** [comment]
> I'm looking into the issue currently, no need to create a new issue, I will create one when needed. Right now I can't reliably reproduce the timestamp update of `package.json` file. With identical setup it sometimes updates and sometimes does not update for my local build. Will keep you posted when I figure out what is going on.

Did you try @steve-todorov's project?
I think the code just needs to compare the hash of the current and new, and only write if they are different.

**Comment by mshabarov** [comment]
Fix has been applied to Vaadin Flow 24.3 (pre) and 24.2, will be released soon for both branches and available to testing.

**Comment by steve-todorov** [comment]
Hey @tepi, @mshabarov ,

We've just upgraded to 24.2.5, but it looks like this does not fix the problem for us.
We are still getting the same error in GH Actions CI:

![image](https://github.com/vaadin/flow/assets/2303663/b977c901-6bea-4635-a096-fdc4d31fd2b6)

Locally it seems to be working as expected. 
Is there any `project path` that you might be caching? For example the `node_modules/.vaadin/vaadin.json` seem to contain an absolute `projectFolder`. If you are caching that on the CI, then that path might be changing depending on which agent the job lands on.

Should we re-open this ticket or open a new one?

**Comment by mcollovati** [comment]
I did some tests both locally and on our gradle starter repository.

Locally, the cache seems to work fine. If I build the project and then do a clean (removing also `frontend/generated` stuff) and rebuild, the log says `:vaadinPrepareFrontend FROM-CACHE`.

If I do not remove the `frontend/generated` folder, I get a non-cachable warning, but not for package.json 
```
Non-cacheable because Gradle does not know how file 'frontend/generated/vaadin.ts' was created (output property 'outputProperties$vaadin_gradle_plugin.generatedTsFolder'). Task output caching requires exclusive access to output paths to guarantee correctness (i.e. multiple tasks are not allowed to produce output in the same location). [OVERLAPPING_OUTPUTS]
```

On GitHub actions I always get the non-cachable warning for `package.json` even if the file is not changed in any way; in the job summary the checksum, creation and modification times are the same before and after the gradle build.

<details>
  <summary>Excerpt of GH build</summary

```
Pre build details
Checksums:

a550b63a7820f031dc93c27124801dbd  package.json
da23cc6f131fce28dfdb79955f59ed53  package-lock.json
File stats

  File: package.json
  Size: 10593     	Blocks: 24         IO Block: 4096   regular file
Device: 801h/2049d	Inode: 283258      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1001/  runner)   Gid: (  127/  docker)
Access: 2023-12-05 10:31:10.656979086 +0000
Modify: 2023-12-05 10:31:06.344961327 +0000
Change: 2023-12-05 10:31:06.344961327 +0000
 Birth: 2023-12-05 10:31:06.344961327 +0000


  File: package-lock.json
  Size: 576628    	Blocks: 1128       IO Block: 4096   regular file
Device: 801h/2049d	Inode: 283257      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1001/  runner)   Gid: (  127/  docker)
Access: 2023-12-05 10:31:10.708979337 +0000
Modify: 2023-12-05 10:31:06.344961327 +0000
Change: 2023-12-05 10:31:06.344961327 +0000
 Birth: 2023-12-05 10:31:06.340961311 +0000

Post build details
Checksums:

a550b63a7820f031dc93c27124801dbd  package.json
da23cc6f131fce28dfdb79955f59ed53  package-lock.json
File stats

  File: package.json
  Size: 10593     	Blocks: 24         IO Block: 4096   regular file
Device: 801h/2049d	Inode: 283258      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1001/  runner)   Gid: (  127/  docker)
Access: 2023-12-05 10:31:10.656979086 +0000
Modify: 2023-12-05 10:31:06.344961327 +0000
Change: 2023-12-05 10:31:06.344961327 +0000
 Birth: 2023-12-05 10:31:06.344961327 +0000


  File: package-lock.json
  Size: 576628    	Blocks: 1128       IO Block: 4096   regular file
Device: 801h/2049d	Inode: 283257      Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1001/  runner)   Gid: (  127/  docker)
Access: 2023-12-05 10:32:20.589286067 +0000
Modify: 2023-12-05 10:32:10.629243784 +0000
Change: 2023-12-05 10:32:10.629243784 +0000
 Birth: 2023-12-05 10:31:06.340961311 +0000
```
</details>

**Comment by mcollovati** [comment]
I'm currently out of ideas. I tried many different setups on GitHub actions by I don't understand why gradle behaves differently from a local run.

The problem might be that the `package.json` file is already there before the task starts, but it is also defined as an output for the task itself. However, locally it does not complain about that.

**Comment by steve-todorov** [comment]
I've added `frontend/generated` into github's action cache to see if that would solve it. Unfortunately it does not.
This problem might be related to NPM doing something behind the scenes to the `package.json` or perhaps there is an environment difference that causes the problem to appear in GH Actions. 

If anybody has time to dig into this you could use a custom self-hosted runner from https://github.com/myoung34/docker-github-actions-runner and try to debug it locally (i.e. log into the container and check what's really going on)

**Comment by mcollovati** [comment]
I was able to reproduce the issue in a local test case.
The issue can also be reproduced locally with a fresh clone of the repository with an already existing `package.json` file.

The problem is that to determine overlapped outputs, gradle compares the current outputs with the ones from the last execution (taken from a project local history). At the first execution, there's obviously not history, so it compares the hash of the current `package.json` file with a default `EMPTY` hash, and this is why it is saying that `Gradle does not know how file 'package.json' was created`.

If in the GH actions workflow, you add a step that executes `gradle vaadinPrepareFronted` before the build, you will see that the new step will not cache anything, whereas the build step will cache the `vaadinPrepareFronted` outputs.

We should check if there's a way to configure the task output/inputs in a way that an existing `package.json` is not considered `overlapping`, or find a different way to handle it.

@mvysny do you perhaps have some suggestions?

**Comment by oliveryasuna** [comment]
Still a problem in Vaadin 24.3.1.

**Comment by benstpierre** [comment]
It would be super amazing to get this fixed.  Shaving 3+ mins off each Github Actions build would be a game changer for my project.

**Comment by steve-todorov** [comment]
Has there been any news regarding this problem? 

Because the `vaadinPrepareFrontend` task is currently not cached in the CI, this also causes the `vaadinBuildFrontend` cache to be invalid.

**Comment by mvysny** [comment]
I think I finally understand what the problem is.

Vaadin plugin is modifying `package.json` in-place - it reads the json file, and if there is a need for modification, it modifies the `package.json` file. Therefore, the property should be annotated both with `@InputFile` and `@OutputFile` but that doesn't work. I've asked a question on SO, hopefully we'll learn more: https://stackoverflow.com/questions/77807234/gradle-cache-task-which-processes-a-file-in-place

Therefore, currently the `package.json` is annotated only with `@OutputFile`. However, that creates a huge build-cache-related problem which is not immediately visible. If the `package.json` file doesn't exist then everything works: the outcome of `vaadinPrepareFrontend` is cached in the build cache and the build cache is reused during the next run of `vaadinPrepareFrontend`. However, if the `package.json` file exists before the build cache is populated, then the following happens:

1. Since `package.json` exists and Gradle doesn't have it in its build cache, it will print the "Gradle does not know how file" blah blah overlapping error message. It will run the `vaadinPrepareFrontend` task fully.
2. However, note the following message which is present too: `Caching disabled for task ':compileJava' because:`. That means that even though `vaadinPrepareFrontend` ran fully, its outcome is not recorded in Gradle's build cache!
3. Next time the build starts, `package.json` exists and Gradle doesn't have it in its build cache. Goto 1.

This behavior is reproducible also with the Java Compiler Gradle task as documented at https://mvysny.github.io/gradle-overlapping-outputs/ .

I'm not sure whether this is a bug in Gradle or whether this is desired behavior; I'll ask on Gradle forums.

**Comment by mvysny** [comment]
There is an answer for the question at https://stackoverflow.com/questions/77807234/gradle-cache-task-which-processes-a-file-in-place . While the answer is correct and the solution would most probably work, it would require us to create synthetic tasks. Those tasks would do nothing and they would only serve the purpose of copying package.json to a subfolder and then back again to the root folder. This solution feels quite complex and hard to understand and maintain. I believe the best way is to create a feature request at https://github.com/gradle/gradle to support this kind of use-case cleanly and correctly.

**Comment by steve-todorov** [comment]
@mvysny thanks for your continuous efforts into debugging and solving this problem! 

Although I see your point of view as valid, I believe waiting for this feature request to be resolved might take a considerable time. Let's open a ticket and see how it goes and if the Gradle team would be open to looking into this.

In the mean time or if Gradle decide to decline fixing this as it is "working as intended" -- it might be more useful to create the necessary additional tasks and properly document why they're needed and how they work. That should make things easier to understand for other team members and potentially easier to maintain.

**Comment by mvysny** [comment]
@steve-todorov Yup, that sounds like a good way forward. I've created a feature request at https://github.com/gradle/gradle/issues/27696  and I agree to continue exactly as you suggested. Let's discuss the ticket with the Gradle guys a bit first; depending on the outcome we can decide appropriate course of action.

**Comment by steve-todorov** [comment]
@mshabarov, @mvysny There has been no response from Gradle yet and it's been more than a month. I believe it would be better if the workaround is implemented and released while we wait for Gradle's team to reply, because at this speed it might take a while longer and CI performance is really hurting our builds. It takes anywhere between 4-5 to 10 minutes for the `vaadinPrepareFrontend` and `vaadinBuildFrontend` to finish even if we are just re-running the same pipeline with no code changes (i.e. only workflow changes).

**Comment by steve-todorov** [comment]
@mvysny  There has been an update from Gradle. It looks like the only way forward would be to implement the workaround.

**Comment by mshabarov** [comment]
@steve-todorov thanks for the update, we will investigate further if the proposed workaround works well.

**Comment by mvysny** [comment]
My two cents - my personal opinion follows.

@steve-todorov if your build takes 5-10 minutes then there's something really wrong going on, and the problem may not lie in the Vaadin plugin. I'd start by trying to profile the Vaadin Gradle plugin: break down the `vaadinPrepareFrontend` and `vaadinBuildFrontend` tasks into individual steps and calculate their durations and log them as DEBUG (please create a new issue for this). Then we'll know what exactly takes long time.

I expect that the npm-related stuff will be the culprit here; once we have proof we can continue with optimizations. The problem is that it's almost impossible to cache `vaadinBuildFrontend` - for example if your npm dependency says 6.2.0+ then the external repository says which dependency will be used. What I'm saying is that `vaadinBuildFrontend` may not even be cacheable since npm is not a pure function.

Therefore, you might need to solve this differently - host a npm repository cache on your premises so that the downloads happen instantly; use pnpm which caches the artifacts locally (then preserve the cache between builds); add more CPU power if the build looks CPU-limited.

In short, it might happen that Vaadin tasks may fundamentally not be cacheable when npm is involved, since npm is not a pure function. Therefore it might be better to start investigating other ways to improve build times.

**Comment by steve-todorov** [comment]
@mvysny Sorry for my delayed reply on this. I've been side tracked by other tasks and only now had a chance to come back to this again. 

I believe I should clarify -- the `vaadinPrepareFrontend` and `vaadinBuildFrontend` tasks are not cached. The `vaadinPrepareFrontend` task is running fast, but the `vaadinBuildFrontend` takes a very long time. Both of these are dependencies of the `jar` task which also takes some significant time in order to produce the jar and then our shadow jar.

If the `vaadinPrepareFrontend` and/or `vaadinBuildFrontend` are not `cached`, then the entire execution graph will be triggered, even if nothing changed.

> My hunch tells me that there's a fundamental problem with Vaadin Gradle plugin task caching. Say one of your npm dependency (may be a transitive one) says 6.2.0+. Then it's up to the external repository says which dependency will be used - that means that npm may produce different results for the same input files. What I'm saying is that vaadinBuildFrontend may not even be cacheable since npm is not a pure function; task being a pure function is a prerequisite for Gradle task caching.

You have a valid point here, although I'd argue transitive dependencies should normally be locked to minor/patch versions which would contain minor fixes. Based on that it might be acceptable to create an arbitrary "cachable output". Maybe something like the maven snapshots where the snapshot is not checked for a new version for the next 24 hours or if anybody has better ideas. Otherwise there would be no way to make this task cachable nor the `vaadinBuildFrontend` which would cause the entire graph to be always executed.

> host a npm repository cache on your premises so that the downloads happen instantly; use pnpm which caches the artifacts locally (then preserve the cache between builds); add more CPU power if the build looks CPU-limited.

We're using Github Actions CI and are caching the npm packages. Still, the build takes a while (even with higher spec builders). Specifically because npm builds are running on a single thread, so it really doesn't matter how much cores you have unless your process has high clock speed.
